Shows a Jetpack connected success text when viewing MC setup and settings pages as a non jetpack owner

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #624 

- Add a util function to get the connected jetpack info
- Show jetpack connected info according to whether the current user is the jetpack owner on the MC setup and Setting pages

### Screenshots:

#### MC setup

![image](https://user-images.githubusercontent.com/17420811/126432895-1d485372-bd5a-4892-bf0f-b25d59a7f4f3.png)

#### Settings

![image](https://user-images.githubusercontent.com/17420811/126433107-a1265cbe-924d-4442-bb17-ae4f4dca25a5.png)

### Detailed test instructions:

1. Log in as the jetpack owner user
2. Go to the MC setup and settings pages to see if it shows the owner email
3. Create a new admin user and log in as that user
4. Go to the MC setup and settings pages to see if it shows "Successfully connected through Jetpack"

### Changelog entry

> Fix - Shows a Jetpack connected success text instead of blank when viewing the onboarding setup and the settings pages as a non jetpack owner account.

eason9487 committed:
Add a util function to get the connected jetpack info

eason9487 committed:
Show jetpack connected info conditionally on the Setting page

eason9487 committed:
Show jetpack connected info conditionally on the MC setup page

ecgan commented on js/src/data/utils.js:
```diff
@@ -167,6 +168,34 @@ export function calculateDelta( value, base ) {
+ * @return {string} Return email if current user is jetpack owner.
+ *                  Otherwise, return a successfully connected info only.
+ */
+export function getConnectedJetpackInfo( jetpack ) {
```
💅  I would put this function in `js/src/utils` instead of `js/src/data/utils`, because its concern is not really on wp-data, but more on what to display for the UI (used in two React components) depending on `jetpack`.

eason9487 replied:
Moved by https://github.com/woocommerce/google-listings-and-ads/pull/893/commits/33d6f5153b6c05df6c39741fad5a118f3bc8d8ae

ecgan commented on js/src/data/utils.test.js:
```diff
@@ -263,3 +264,34 @@ describe( 'mapReportFieldsToPerformance', () => {
+			const info = getConnectedJetpackInfo( jetpack );
+
+			expect( info ).not.toBe( jetpack.email );
+			expect( info ).toMatchSnapshot();
```
❓  Since the phrase `"Successfully connected through Jetpack"` is a requirement as defined by Jill in https://github.com/woocommerce/google-listings-and-ads/issues/624#issuecomment-883829052, I would code it into the test. One `expect` instead of two, and we don't need the `js/src/data/__snapshots__/utils.test.js.snap` file. What do you think?

```suggestion
			expect( info ).toBe( 'Successfully connected through Jetpack' );
```

eason9487 replied:
The main concern of this test is to verify whether the function makes logical condition based on `owner`, and the returned textual content is not the primary target of this test. Using snapshot can still verify whether the text content is correct. When the text change happened, the snapshot could alert developers as well and could be easily updated if the change is confirmed. Furthermore, it doesn't need to maintain another text copy in test case.

> One `expect` instead of two, and we don't need the `js/src/data/__snapshots__/utils.test.js.snap` file. What do you think?

However, using only the snapshot to verify the condition would be a bit ambiguous in expressing the test intention, so an additional `expect( info ).not.toBe( jetpack.email );` is used to enhance the expressiveness. Just like the `checkErrors` test, where the snapshot is verified along with a `expect( errors ).toHaveProperty( 'shipping_rate' );` to strengthen the expression of test intention.

https://github.com/woocommerce/google-listings-and-ads/blob/20247e60a491f2546bcaee5862d30f678b01557f/js/src/components/free-listings/configure-product-listings/checkErrors.test.js#L68-L73

---

I have to say that this test case is not precise enough, either with snapshot or with `.toBe()` assertion. When I developed it, I thought the `getConnectedJetpackInfo` function was so straightforward that it was almost impossible to get it wrong, so I'm probably okay with a little less precision.

But to make it precise, then I would change it to a pure function like this

```js
function getConnectedJetpackInfo( jetpack, nonOwnerInfo ) {
	// ... omitted

	if ( jetpack.owner === 'yes' ) {
		return jetpack.email;
	}

	// No available owner info. Show `nonOwnerInfo` instead.
	return nonOwnerInfo;
}
```

And assert it by

```js
it( 'When current user is not the jetpack owner, should return `nonOwnerInfo` param instead', () => {
	const jetpack = {
		active: 'yes',
		owner: 'no',
		email: 'i_am_another_admin_user@example.com',
	};
	const nonOwnerInfo = {}; // Use object to assert the same reference

	const info = getConnectedJetpackInfo( jetpack, nonOwnerInfo );

	expect( info ).toBe( nonOwnerInfo );
} );
```

Based on the premise of a very simple function, that would be a bit overkill for me and need more codes when used. But it ensures the accuracy of the test. So, in terms of overall balance, I chose the current implementation when I was developing it. Which would you prefer?

ecgan replied:
> ... to verify whether the function makes logical condition based on `owner`, and the returned textual content is not the primary target of this test.

When I first see the function and its JSDoc, I thought the responsibility of the function is to make the logical condition, plus having the string `'Successfully connected through Jetpack'` in one place, so that other places calling the function will get the same string. That's why I thought it make sense to test for the string. What do you think on this?

> Furthermore, it doesn't need to maintain another text copy in test case.

Having the text copy in test case is common and expected, right? I mean, if there is a requirement that says "the component / function / application should display 'Hello World'", the test for it would need to assert the existence of the 'Hello World' text.

> Which would you prefer?

The one you presented here:

```
	const info = getConnectedJetpackInfo( jetpack, nonOwnerInfo );
```

... yeah, it does look overkill. Also, it would require the `nonOwnerInfo` (i.e. the string `'Successfully connected through Jetpack'`) to be placed elsewhere and to be passed into the function call, which feels like not so ideal in terms of developer experience (which is why I thought good to put the string in the function).

So comparing the code you have in this PR and the one you presented above, the one in this PR in good. 😄

eason9487 replied:
> That's why I thought it make sense to test for the string. What do you think on this?

For this case, the wording of expression is not the main focus of the test. I thought snapshot is an appropriate usage, and it will test whether the string is matched as well.

There are many similar contexts in [checkErrors.test.js](https://github.com/woocommerce/google-listings-and-ads/blob/20247e60a491f2546bcaee5862d30f678b01557f/js/src/components/free-listings/configure-product-listings/checkErrors.test.js). After viewing every `toMatchSnapshot()` usages, maybe it could show more sense why I adopted this method. But If devs thought that every time they read these tests, they have to open the function implementation or snapshot files to confirm what the string exactly is, it really affects the developer experience. Then I would be happy to open another PR to have these strings within test cases directly.

> Having the text copy in test case is common and expected, right?

If the returned string of function is a constant, protocol, code, or something like that, having the string copy within test case is common and expected. For example:

```js
function detectSocialNetwork( url ) {
	if ( matchTwitter( url ) ) {
		return 'twitter';
	}
	if ( matchFacebook( url ) ) {
		return 'facebook';
	}
	if ( matchInstagram( url ) ) {
		return 'instagram';
	}
	return null;
}
```

> I mean, if there is a requirement that says "the component / function / application should display 'Hello World'", the test for it would need to assert the existence of the 'Hello World' text.

I think it depends on the text is actually displayed by a component, function, or application. And how the "display" is to be defined. It would bring different practices.

ecgan commented on js/src/setup-mc/setup-stepper/setup-accounts/wordpressdotcom-account/card-content.js:
```diff
@@ -30,7 +31,8 @@ const CardContent = () => {
 	if ( jetpack.active === 'yes' ) {
-		return <TitleButtonLayout title={ jetpack.email } />;
+		const title = getConnectedJetpackInfo( jetpack );
+		return <TitleButtonLayout title={ title } />;
```
💅 😄  very small nit, we can just squeeze it like this, similar to `DisconnectAccounts`.

```suggestion
		return (
			<TitleButtonLayout title={ getConnectedJetpackInfo( jetpack ) } />
		);
```

ecgan commented on js/src/data/utils.test.js:
```diff
@@ -263,3 +264,34 @@ describe( 'mapReportFieldsToPerformance', () => {
 } );
+
+describe( 'getConnectedJetpackInfo', () => {
+	it( 'When jetpack in not connected, should return an empty string', () => {
+		expect( getConnectedJetpackInfo( { active: 'no' } ) ).toBe( '' );
+	} );
```
💅 ❔ Just wondering whether we should use a `describe` block for this so that it is consistent with the other `describe` block.

```suggestion
	describe( 'For non-connected jetpack', () => {
		it( 'should return an empty string', () => {
			expect( getConnectedJetpackInfo( { active: 'no' } ) ).toBe( '' );
		} );
	} );
```

eason9487 replied:
I would say both works for me. And `describe` would be more meaningful to group the related test cases if they have a commonality in the purpose or premise. However, based on the consistency of the typography or code blocks, it is not a priority when writing tests, especially since one more layer group has only one case.

The same blocks structure:

https://github.com/woocommerce/google-listings-and-ads/blob/20247e60a491f2546bcaee5862d30f678b01557f/js/src/reports/utils.test.js#L17-L33

ecgan reviewed (commented):
Tested as per testing instruction, works as expected. Also tested with `npm run test-unit`, all tests passed. 

Left one ❓ comment, and a few other 💅 ❔  comments.

eason9487 committed:
Squeeze the calling of getConnectedJetpackInfo into TitleButtonLayout

Address https://github.com/woocommerce/google-listings-and-ads/pull/893#discussion_r674043695

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

eason9487 committed:
Move getConnectedJetpackInfo to .~/utils/

Address https://github.com/woocommerce/google-listings-and-ads/pull/893#discussion_r674034773

ecgan reviewed (approved):
Approving this, since @eason9487 has already got back to me on the comments. 

Regarding https://github.com/woocommerce/google-listings-and-ads/pull/893#discussion_r674871897, if you like to make changes for it, go ahead; otherwise, since I have already got your perspective on it, feel free to merge it. 🙂
